contracts/Challenge1.2.sol

```diff
@@ -0,0 +1,105 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "./IUniswapV2Router02.sol";
+import "./IWETH.sol";
+import "./IterableMap.sol";
+
+contract SwapperUNIV2 is ReentrancyGuard {
+
+    using IterableMapping for IterableMapping.Map;
+
+    address public DAI;
+    address public WETH;
+    IUniswapV2Router02 public UniswapV2Router;
+    IterableMapping.Map private DAIBalances;
+    IterableMapping.Map private WETHBalances;
+
+    constructor (address _DAI, address _WETH, address _UniswapV2Router) {
+        DAI = _DAI;
+        WETH = _WETH;
+        UniswapV2Router = IUniswapV2Router02(_UniswapV2Router);
+        //Unilimited approve
+        IERC20(WETH).approve(_UniswapV2Router, type(uint).max);
+    }
+
+    function _swap(uint256 relPrice) internal {
+        //Swap all tokens
+        for (uint i = 0; i < WETHBalances.size(); i++) {
+            address key = WETHBalances.getKeyAtIndex(i);
+            uint256 balanceWETH = WETHBalances.get(key);
+
+            if (balanceWETH > 0) {
+                WETHBalances.set(key, 0);
+                uint256 balanceDAI = DAIBalances.get(key);
+                DAIBalances.set(key, balanceDAI + (balanceWETH * relPrice));
+            }
+        }
+    }
+
+    function _getAllWETHBalance() internal view returns (uint256) {
+        uint256 count = 0;
+        for (uint i = 0; i < WETHBalances.size(); i++) {
+            address key = WETHBalances.getKeyAtIndex(i);
+            uint256 balance = WETHBalances.get(key);
+            count += balance;
+        }
+        return count;
+    }
+
+    function provide(uint256 amount, bool isNativeETH) external payable {
```

Comment:
You could create two overloaded functions, for example:
`function provide(uint256 amount) external`
and
`function provide() external payable`